viewing_party/party.py

```diff
@@ -1,23 +1,266 @@
+
+#     return final_rec_list
+
+# # ------------- WAVE 5 --------------------from HP
+
+# def get_new_rec_by_genre(user_data):
+
+#     if not user_data["watched"]:
+#         return []
+
+#     most_seen_genre = get_most_watched_genre(user_data)
+
+#     watched_list = user_data["watched"]
+#     watched_titles = {movie["title"] for movie in watched_list}
+
+#     added_titles = set()
+#     final_rec_list = []
+
+#     for friend in user_data["friends"] :
+#         for movie in friend["watched"] :
+#             if movie["genre"] == most_seen_genre :
+#                 if movie["title"] not in watched_titles:
+#                     if movie["title"] not in added_titles:
+#                         final_rec_list.append(movie)
+#                         added_titles.add(movie["title"])
+
+#     return final_rec_list
+
+
+# def get_rec_from_favorites(user_data):
+#     if not user_data or not user_data["favorites"]:
+#         return []
+
+#     recommended_movies = []
+
+#     for favorite_movie in user_data["favorites"]:
+#         found = False
+#         for friend in user_data["friends"]:
+#             for friends_watched_movie in friend["watched"]:
+#                 if favorite_movie == friends_watched_movie:
+#                     found = True
+#                     break
+#             if found:
+#                 break
+
+#         if not found:
+#             recommended_movies.append(favorite_movie)
+
+#     return recommended_movies
+
+# ------------- WAVE 4 --------------------from AT
+
+def get_available_recs(user_data):
+    recommended_movies = []
+    current_subscriptions = set(user_data["subscriptions"])
+
+    friends_watched_movies = get_friends_unique_watched(user_data)
+
+    for movie in friends_watched_movies:
+        if movie["host"] in current_subscriptions:
+            recommended_movies.append(movie)
+
+    return recommended_movies
+
+
+# ------------- WAVE 5 --------------------from AT
+
+def get_new_rec_by_genre(user_data):
+    recommended_movies = []
+    genre_count = {}
+    for movie in user_data["watched"]:
+        if movie["genre"] not in genre_count:
+            genre_count[movie["genre"]] = 0
+
+        genre_count[movie["genre"]] += 1
+
+    fav_genre = ""
+    max_genre_count = -1
+
+    for genre, count in genre_count.items():
+        if max_genre_count < count:
+            fav_genre = genre
+            max_genre_count = count
+
+    friend_movies = get_friends_unique_watched(user_data)
+
+    for movie in friend_movies:
+        if movie["genre"] == fav_genre:
+            recommended_movies.append(movie)
+
+    return recommended_movies
+
+def get_rec_from_favorites(user_data):
+    recommended_movies = []
 
-# -----------------------------------------
-# ------------- WAVE 2 --------------------
-# -----------------------------------------
+    favorite_movies = {}
 
+    for movie in user_data["favorites"]:
+        favorite_movies[movie["title"]] = movie

```

Comment:
Smart approach to make the favorite movies list into a dictionary! One small tweak I would make would be to use a set rather than a dictionary. Later on when we loop through our movies, we already have all the data associated with each movie as we loop through them, so we don't need to hold onto all of the data here as well. We could easily get away with just holding onto the favorite titles with a set!